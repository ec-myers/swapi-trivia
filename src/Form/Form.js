import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Form.scss';

class Form extends Component {
  constructor(getFormData) {
    super(getFormData);
    this.state = {
      name: "",
      quote: "",
      rank:"Padawan",
      nameErr: false,
      quoteErr: false,
      isComplete: false
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.id]: e.target.value})
  }

  handleSubmit = (e) => {
    e.preventDefault()
    if (!this.state.name) {
      this.setState({nameErr: true})
    } else { this.setState({ nameErr: false }) }

    if (!this.state.quote) {
      this.setState({quoteErr: true})
    } else { this.setState({ quoteErr: false })}

    if(this.state.name && this.state.quote) {
      let user = {name:this.state.name, 
        quote:this.state.quote, 
        rank: this.state.rank}
      this.props.getFormData(user)
      this.setState({isComplete: true})
    }
  }

  render() {
    if (this.state.isComplete) {
      return <Redirect to='/movies' />
    }

    const { name, quote, rank, nameErr, quoteErr } = this.state
    let nameClass = nameErr ? "error" : "";
    let quoteClass = quoteErr ? "error" : "";


    return (
      <form>
        <div className="form-background">
          <label htmlFor="name">Name</label>
          <input autoFocus value={name} id="name" 
          placeholder="Enter a name"
          className={nameClass}
          onChange={this.handleChange}/>
          <div>
          {nameErr && <p>Please enter a name</p>}
          </div>
          <label htmlFor="quote">Favorite Star Wars Quote</label>
          <input value={quote} id="quote"
          className={quoteClass} 
            placeholder="Enter your favorite Quote"  
          onChange={this.handleChange}/>
          <div>
          {quoteErr && <p>Please enter a quote</p>}
          </div>
          <label htmlFor="rank">Rank</label>
          <select value={rank} 
          onChange={this.handleChange} id="rank">
            <option value="Padawan" selected>Padawan</option>
            <option value="Jedi Knight">Jedi Knight</option>
            <option value="Jedi Master">Jedi Master</option>
          </select>
            <button type="button" onClick={this.handleSubmit}>May The Force Be With You</button>
        </div>
      </form>
    )
  }
}

export default Form;

Form.propTypes = {
  getFormData: PropTypes.func.isRequired
}