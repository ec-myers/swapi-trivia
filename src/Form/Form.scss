form {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100vh;
  input {
    margin: 10px;
    width: 20%;
    padding: 10px;
  }

  select {
    height: 3%;
    margin: 10px;
    width: 20%;
  }

  div {
    min-height: 40px;
    p {
      display: inline;
    }
  }

    button {
      height: 5%;

      margin-top: 10px;
    &:hover {
      cursor: pointer;
    }
  }
  .error {
    outline: 3px solid red;
  }
}

