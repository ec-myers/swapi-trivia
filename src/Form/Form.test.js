import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';

describe('Form', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Form getFormData={jest.fn()}/>)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('should update state when handleChange is called', () => {
    const mockEvent = { target: { id: 'name', value: 'Kimberly' }, preventDefault: jest.fn() };
    const expected = 'Kimberly';
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('name')).toEqual(expected);
  })
  it('should update state to complete when handleSubmit is called', () => {
    const mockEventOne = { target: { id: 'name', value: 'Kimberly' }}
    const mockEventTwo = { target: {
      id: 'quote', value: 'I know'
    }}

    const dEvent = { preventDefault:jest.fn()}

    wrapper.instance().handleChange(mockEventOne)
    expect(wrapper.state('name')).toEqual('Kimberly')
    wrapper.instance().handleChange(mockEventTwo)
    expect(wrapper.state('quote')).toEqual('I know')
    wrapper.find('button').simulate('click', dEvent)
    expect(wrapper.state('isComplete')).toEqual(true)
  })
  it('should not be marked complete if a section is missing on submit and should have an error', () => {
    const mockEvent = { target: { id: 'name', value: 'Kimberly' }, preventDefault: jest.fn() };
    const expected = 'Kimberly';
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('name')).toEqual(expected);
    expect(wrapper.state('quote')).toEqual('')
    wrapper.find('button').simulate('click', mockEvent)
    expect(wrapper.state('isComplete')).toEqual(false)
    expect(wrapper.state('quoteErr')).toEqual(true)
  })
  it('should run handleChange when the inputs detect a change', () => {
    const mockNameEvent = { target: { id: 'name', value: 'Robbie' } };
    const mockQuoteEvent = { target: { id: 'quote', value: 'Why not?' } };
    wrapper.instance().handleChange = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('[id="name"]').simulate('change', mockNameEvent);
    wrapper.find('[id="quote"]').simulate('change', mockQuoteEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockNameEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockQuoteEvent);
  }) 
  it('should call getFormData with correct props when the button is clicked', () => {
    const currentState = { name: 'Stephanie', quote: 'I have a bad feeling about this', rank: 'Jedi Master' };
    let mockUser = {
      name: 'Stephanie',
      quote: 'I have a bad feeling about this',
      rank: 'Jedi Master'
    }
    wrapper.instance().setState(currentState);
    wrapper.instance().handleSubmit({preventDefault: jest.fn() });
    expect(wrapper.instance().props.getFormData).toHaveBeenCalledWith(mockUser);
  })
})