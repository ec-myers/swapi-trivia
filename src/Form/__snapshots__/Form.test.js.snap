// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
<form>
  <div
    className="form-background"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      autoFocus={true}
      className=""
      id="name"
      onChange={[Function]}
      placeholder="Enter a name"
      value=""
    />
    <div />
    <label
      htmlFor="quote"
    >
      Favorite Star Wars Quote
    </label>
    <input
      className=""
      id="quote"
      onChange={[Function]}
      placeholder="Enter your favorite Quote"
      value=""
    />
    <div />
    <label
      htmlFor="rank"
    >
      Rank
    </label>
    <select
      id="rank"
      onChange={[Function]}
      value="Padawan"
    >
      <option
        selected={true}
        value="Padawan"
      >
        Padawan
      </option>
      <option
        value="Jedi Knight"
      >
        Jedi Knight
      </option>
      <option
        value="Jedi Master"
      >
        Jedi Master
      </option>
    </select>
    <button
      onClick={[Function]}
      type="button"
    >
      May The Force Be With You
    </button>
  </div>
</form>
`;
